parts/k8s/cloud-init/masternodecustomdata.yml

```diff
@@ -529,6 +546,85 @@ MASTER_CONTAINER_ADDONS_PLACEHOLDER
     {{WrapAsVariable "environmentJSON"}}
 {{end}}
 
+{{- if .MasterProfile.IsFlatcar}}
+- path: /opt/azure/containers/provision-setup.sh
+  permissions: "0755"
+  owner: root
+  content: |
+    #!/bin/bash
+    source {{GetCSEHelpersScriptFilepath}}
+    /opt/azure/containers/mountetcd.sh
+    retrycmd_if_failure 5 5 10 curl --retry 5 --retry-delay 10 --retry-max-time 10 --max-time 60 https://127.0.0.1:2379/v2/machines
+    {{- if EnableAggregatedAPIs}}
+    sudo bash /etc/kubernetes/generate-proxy-certs.sh
+    {{- end}}
+    touch /opt/azure/containers/runcmd.complete
+
+- path: "/etc/kubernetes/manifests/.keep"
+
+  {{- if .OrchestratorProfile.KubernetesConfig.RequiresDocker}}
+groups:
+  - docker: [{{WrapAsParameter "linuxAdminUsername"}}]
+  {{- end}}
+
+# This must be coreos, not flatcar until (put reference link here)
+# hits all the channels.
+coreos:
+  units:
+    - name: start-provision-setup.service
+      command: "start"
+      content: |
+        [Unit]
+        Description=Start provision setup service
+
+        [Service]
+        ExecStart=/opt/azure/containers/provision-setup.sh
+    - name: kubelet.service
+      enable: true
+      drop-ins:
+        - name: "10-flatcar.conf"
+          content: |
+            [Unit]
+            Requires=rpc-statd.service
+            After=etcd.service
+            ConditionPathExists=
+            ConditionPathExists=/opt/kubelet
+            [Service]
+            ExecStart=
+            ExecStart=/opt/kubelet \
+              --enable-server \
+              --node-labels="${KUBELET_NODE_LABELS}" \
+              --v=2 \
+              --volume-plugin-dir=/etc/kubernetes/volumeplugins \
+              $KUBELET_CONFIG $KUBELET_OPTS \
+              $KUBELET_REGISTER_NODE $KUBELET_REGISTER_WITH_TAINTS
+    - name: kubelet-monitor.service
+      enable: true
+      drop-ins:
+        - name: "10-flatcar.conf"
+          content: |
+            [Service]
+            ExecStart=
+            ExecStart=/opt/bin/health-monitor.sh kubelet
+    - name: docker-monitor.service
+      enable: true
+      drop-ins:
+        - name: "10-flatcar.conf"
+          content: |
+            [Service]
+            ExecStart=
+            ExecStart=/opt/bin/health-monitor.sh container-runtime
+    - name: etcd.service
+      enable: true
+      drop-ins:
+        - name: "10-flatcar.conf"
+          content: |
+            [Service]
+            ExecStart=
+            ExecStart=/opt/bin/etcd $DAEMON_ARGS
+    - name: etcd-member.service
+      mask: true
+{{else}}
 disk_setup:
```

Comment:
@johnsonshi it looks like flatcar doesn't support these cloud-init interfaces? How should we be mounting `/var/lib/etcddisk` on flatcar-enabled master VMs?

Alternatively, we can only add flatcar support as a worker node, and require Ubuntu for master nodes.

cc @rbitia @ritazh

Reply:
@invidian can you confirm that it makes sense to scope this work to worker nodes only? I can fix up the PR so that flatcar is enabled as a node distro, but not available for the master (control plane) VMs.

Reply:
The other alternative would be for Flatcar to include those cloud-init udev rules so that the symlinks are created. However that may exceed this PR's scope.

Reply:
Waagent also has the udev rules that creates symlinks to data disks at /dev/disk/azure/scsi/lun#.

I deployed a Flatcar VM. I confirmed that the udev rules for data disks are present. I then used `udevadm monitor` to monitor for udev events, then used a separate console to attach a data disk. From the monitor I see that the udev rule is fired. After that, I see that the symlink has been created.

https://paste.ubuntu.com/p/rqCqh7BPcT/

Reply:
From my tests, Flatcar was working just fine for master nodes. I didn't test the reboots though, which were problematic in the past :thinking:

Reply:
Are there any known issues for Flatcar for master nodes? I can investigate that, if you point me to them.

We could only merge support for worker nodes, but I'd rather avoid doing more changes to this, already large changeset if possible, I'd rather prefer incremental improvements and fixes. The whole reverting was a lot of work and resolving conflicts while rebasing on `master` is also not trivial.

Reply:
Hey @invidian I think @jackfrancis 's concern is whether Flatcar can correctly mount the etcd data disk. Being able to mount the etcd data disk is a prerequisite for a distro image to become a master node.

Well from my validations (and from a cloud-init provisioning perspective), Flatcar can correctly mount the etcd data disk. Here's why:

The WALinuxAgent package comes with the required [udev rules](https://github.com/Azure/WALinuxAgent/blob/develop/config/66-azure-storage.rules) for handling attached VM data disks. These udev rules are required on Azure VMs to create publicly-documented symlinks from `/dev/disk/azure/scsi1/lun#` to the attached VM data disks. From [this specific line in the WALinuxAgent udev rules](https://github.com/Azure/WALinuxAgent/blob/78c660062f2ef86976cf4d03c8d99b70fe0a7255/config/66-azure-storage.rules#L12), we can see that the udev rule creates symlinks from `/dev/disk/azure/scsi1/lun#` to VM data disks (which will always have `f8b3781b-1e82-4818-a1c3-63d806ec15bb` as the device ID`). Since the WALinuxAgent package comes with these rules and since the package is installed by default in Flatcar, these udev rules are guaranteed to always be in the system.

Well I tested to see if Flatcar is able to handle a VM data disk (such as the etcd data disk). I did my validation over [here](https://paste.ubuntu.com/p/rqCqh7BPcT/). From that link (my manual validation), you can see that `/lib/udev/rules.d/66-azure-storage.rules` contains the WALinuxAgent udev rules. Furthermore, I used `udevadm monitor` to monitor for kernel udev events. I then attached a VM data disk to the Flatcar VM from another console, and as expected, the udev rules were fired, and a symlink from `/dev/disk/azure/scsi1/lun0` is created by the fired udev rule. This symlink allows Flatcar to have a reliable mechanism to refer to the etcd data disk.

Here is the reason why we need the udev rule and symlink to exist on Azure VMs for data disks. On Azure, there is no guarantee after VM reboot that `/dev/sdc` is the etcd data disk. Now, because the udev rule is always present for all Flatcar VMs on Azure, we have a safe guarantee that `/etc/fstab` can have a reliable path (symlink) to refer to the etcd data disk.

If you are keen on reading up on more reasons, here we gooo... :)

Take a look at the [master node cloud-init custom-data YAML at these lines](https://github.com/Azure/aks-engine/blob/2062d0793f58fda7c6ceeb61a3f1dd5004e94ee4/parts/k8s/cloud-init/masternodecustomdata.yml#L538-L543). You can see that the cloud-init `fs_setup` module uses `/dev/disk/azure/scsi1/lun0` to refer to the etcd data disk. Those highlighted lines of the cloud-init `fs_setup` module indicate that the etcd data disk should be formatted to `ext4` and most importantly should have the `etcd_disk` label. Cloud-init happily formats and applies the `etcd_disk` label as instructed.

Now, take a look at [the next few lines](https://github.com/Azure/aks-engine/blob/2062d0793f58fda7c6ceeb61a3f1dd5004e94ee4/parts/k8s/cloud-init/masternodecustomdata.yml#L550-L552). From the [docs on cloud-init's `mounts` module](https://cloudinit.readthedocs.io/en/latest/topics/modules.html?highlight=label#mounts), we see that it says the `mounts` module writes the mount directives to `/etc/fstab`, which ensures that these directives are persisted across reboots. The mount directive (which is written to fstab) instructs the system to "please mount the filesystem with the `etcd_disk` label at `/var/lib/etcddisk`.

Now because the etcd file system has been mounted at `/var/lib/etcddisk`, this triggers the systemd unit [etcd.service](https://github.com/Azure/aks-engine/blob/2062d0793f58fda7c6ceeb61a3f1dd5004e94ee4/parts/k8s/cloud-init/artifacts/etcd.service#L7), which starts etcd. Notice that this systemd unit `RequiresMountsFor=/var/lib/etcddisk`.

Reply:
@invidian @jackfrancis Now the only validations I have run is from a Linux system provisioning perspective, which is specifically validating whether the udev rule is fired and the data disk symlink is created (which both did). Because the symlink is created, the rest of the pieces (cloud-init fs setup and label application, mounting the disk at the expected location, and triggering the etcd systemd service) should logically follow. From a Linux system provisioning perspective, this is a definite pass and LGTM.

However, @invidian mentioned that Flatcar reboots were problematic in the past. I don't know what these problematic issues were, so maybe you can clarify what they were?

Before this thing is merged in, can one or both of you test that (1) Flatcar works without any issues when provisioned as master nodes using this PR's code, and (2) the Flatcar master nodes can be restated with no issues, and (3) after restart the etcd service is up and alive because the etcd disk was mounted at the expected location to trigger the etcd systemd unit.

Reply:
This is why the mounting is not working, from oem-cloudinit logs:

```
Jun 04 23:25:23 k8s-master-36217359-0 coreos-cloudinit[871]: 2020/06/04 23:25:23 line 422: warning: unrecognized key "disk_setup"
Jun 04 23:25:23 k8s-master-36217359-0 coreos-cloudinit[871]: 2020/06/04 23:25:23 line 427: warning: unrecognized key "fs_setup"
Jun 04 23:25:23 k8s-master-36217359-0 coreos-cloudinit[871]: 2020/06/04 23:25:23 line 438: warning: unrecognized key "mounts"
Jun 04 23:25:23 k8s-master-36217359-0 coreos-cloudinit[871]: 2020/06/04 23:25:23 line 362: warning: unrecognized key "groups"
```

Reply:
@jackfrancis and I got together and diagnosed the issue preventing Flatcar from being provisioned correctly for master nodes.

When the Flatcar VM is provisioned as a master node VM with the etcd data disk, the udev rule is indeed fired and the symlink from `/dev/disk/azure/scsi1/lun0` is created. However, as seen in the coreos-cloudinit logs, the issue is that coreos-cloudinit does not recognize [these cloud-init directives](https://github.com/Azure/aks-engine/blob/94cfb92ccc122048f910b61cb5d1c6b2652a9537/parts/k8s/cloud-init/masternodecustomdata.yml#L532-L552).

The coreos-cloudinit codebase makes no mention of these upstream cloud-init directives. Lack of support and significant divergence between coreos-cloudinit and upstream cloud-init is preventing the etcd data disk from being set up correctly.

From a code maintainability point of view, it will be difficult to maintain aks-engine code to support VM provisioning using both upstream cloud-init and coreos-cloudinit. Perhaps Flatcar for worker nodes only is the way to go.

Reply:
Oh, it worked by the time I submitted the PR, but febebda6a6883cf8a1ad4a91ef0f3fe6d9a1a20b changed the way etcddisk is mounted :disappointed:

Reply:
Thanks for explanation @jackfrancis. I understand the reasoning behind this change and it seem indeed reasonable. It's just unfortunate that it happen and this time and that Flatcar does not support required options. 

I think it's fine if we add support only for workers for now and I will collaborate with Flatcar team to maybe add those missing options to Flatcar itself.

Who should revert the changes done in this PR for master nodes? Also, is there a way to keep this revert as a separate commit? Once Flatcar has support for missing options, it would be easier to add support for master nodes then.

Reply:
I'm happy to fix up this PR to make it worker node-only.

I'll file a backlog item that documents that flatcar isn't yet implemented for control plane VMs, and will reference your original PR, which is probably the cleanest changeset representation of how we may want to revisit that work in the future.

How does that sound?

Reply:
Sounds good to me, thank you for your help!